core/src/main/java/com/expedia/adaptivealerting/core/detector/PewmaOutlierDetector.java

```diff
@@ -0,0 +1,186 @@
+import com.expedia.adaptivealerting.core.OutlierLevel;
+
+import java.time.Instant;
+
+import static com.expedia.adaptivealerting.core.util.AssertUtil.isBetweenExclusive;
+import static java.lang.Math.abs;
+import static java.lang.Math.exp;
+import static java.lang.Math.pow;
+import static java.lang.Math.sqrt;
+
+/**
+ * <p>
+ * Anomaly detector based on the probabilistic exponential weighted moving average. This is an online algorithm, meaning
+ * that it updates the thresholds incrementally as new data comes in, but is less influenced by outliers than EWMA.
+ * </p>
+ * <p>
+ * It takes a little while before the internal mean and variance estimates converge to something that makes sense. As a
+ * rule of thumb, feed the detector 30 data points or so before using it for actual anomaly detection.
+ * </p>
+ * <p>
+ * See https://www.ll.mit.edu/mission/cybersec/publications/publication-files/full_papers/2012_08_05_Carter_IEEESSP_FP.pdf.
+ * Implementation based off code from here https://aws.amazon.com/blogs/iot/anomaly-detection-using-aws-iot-and-aws-lambda/.
+ *
+ * </p>
+ *
+ * @author David Sutherland
+ */
+public class PewmaOutlierDetector implements OutlierDetector {
+
+    /**
+     * Smoothing param.
+     */
+    private final double alpha0;
+
+
+    /**
+     * Outlier weighting param.
+     */
+    private final double beta;
+
+    /**
+     * How many iterations to train for.
+     */
+    private final double trainingLength;
+
+    /**
+     * Smoothing param.
+     */
+    private double trainingCount;
+
+    /**
+     * Internal state for PEWMA algorithm.
+     */
+    private double s1;
+
+    /**
+     * Internal state for PEWMA algorithm.
+     */
+    private double s2;
+
+    /**
+     * Weak outlier threshold, in sigmas.
+     */
+    private final double weakThreshold;
+
+    /**
+     * Strong outlier threshold, in sigmas.
+     */
+    private final double strongThreshold;
+
+    /**
+     * Local mean estimate.
+     */
+    private double mean;
+
+    /**
+     * Local standard deviation estimate.
+     */
+    private double stdDev;
+
+    /**
+     * Creates a new PEWMA anomaly detector. Initial mean is given by initValue and initial standard deviation is 0.
+     *
+     * @param initialAlpha    Smoothing parameter.
+     * @param beta            Outlier weighting parameter.
+     * @param trainingLength  How many iterations to train for.
+     * @param weakThreshold   Weak outlier threshold, in sigmas.
+     * @param strongThreshold Strong outlier threshold, in sigmas.
+     * @param initValue       Initial observation, used to set the first mean estimate.
+     */
+    public PewmaOutlierDetector(
+            double initialAlpha,
+            double beta,
+            int trainingLength,
+            double weakThreshold,
+            double strongThreshold,
+            double initValue
+    ) {
+        isBetweenExclusive(initialAlpha, 0.0, 1.0, "initialAlpha must be in the range (0, 1)");
```

Comment:
My reading of the paper had this as exclusive. Not sure if this should also apply to the alpha in EWMA.

Reply:
You are right that the paper treats it as exclusive. I also share your intuition that we should use the same bounds for EWMA and PEWMA since the models are very closely related.

In this case I think we can just choose how we want to do it, since alpha doesn't appear as a denominator or anything like that. Other treatments of EWMA use different bounds; see e.g., https://www.itl.nist.gov/div898/handbook/pmc/section3/pmc324.htm, where the interval is half-open.

Personally I think we should use the closed interval [0, 1] in both cases. Obviously 0 and 1 are degenerate cases but I don't see any reason to disallow totally ignoring history or totally ignoring currents. I don't feel strongly.

Reply:
Fine with me. I'll change it.